src/diagnostic/mod.rs

```diff
@@ -1,4 +1,295 @@
+
+        self
+    }
+
+    /// Adds a note message with a separate span to the diagnostic
+    pub fn span_note(&mut self, span: Span, message: impl Into<String>) -> &mut Self {
+        let subd = SubDiagnostic::new(Level::Note, message.into(), Some(span));
+        self.diagnostic.children.push(subd);
+
+        self
+    }
+
+    /// Adds a help message to the diagnostic
+    pub fn help(&mut self, message: impl Into<String>) -> &mut Self {
+        let subd = SubDiagnostic::new(Level::Help, message.into(), None);
+        self.diagnostic.children.push(subd);
+
+        self
+    }
+
+    /// Adds a help message with a separate span to the diagnostic
+    pub fn span_help(&mut self, span: Span, message: impl Into<String>) -> &mut Self {
+        let subd = SubDiagnostic::new(Level::Help, message.into(), Some(span));
+        self.diagnostic.children.push(subd);
+
+        self
+    }
+
+    /// Queues this diagnostic to be emitted by the inner Handler/Emitter
+    pub fn emit(&mut self) {
+        if self.diagnostic.level == Level::Warning {
+            self.handler.warn(self.diagnostic.clone());
+        } else {
+            self.handler.error(self.diagnostic.clone());
+        }
+
+        // Mark this as cancelled so that it can be safely dropped
+        self.cancel();
+    }
+
+    /// Sets this DiagnosticBuilder as cancelled, meaning that it is safe to be dropped
+    pub fn cancel(&mut self) {
+        self.diagnostic.level = Level::Cancelled;
+    }
+
+    /// Returns true if this was cancelled, false otherwise
+    pub fn cancelled(&self) -> bool {
+        self.diagnostic.level == Level::Cancelled
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Diagnostic {
+    pub level: Level,
+    pub message: String,
+    pub primary: Option<Span>,
+    pub spans: Vec<(Span, String)>,
+    pub children: Vec<SubDiagnostic>,
+}
+
+impl<'a> Drop for DiagnosticBuilder<'a> {
+    fn drop(&mut self) {
+        // DiagnosticBuilders are sort of bombs if dropped. This had better either be emitted, or
+        // cancelled. If not, we emit a bug error.
+        if !self.cancelled() {
+            let mut db = DiagnosticBuilder::new(
+                self.handler,
+                Level::Bug,
+                "the following error was constructed but not emitted".to_string(),
+            );
+
+            db.emit();
+            self.emit();
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SubDiagnostic {
+    pub level: Level,
+    pub message: String,
+    pub span: Option<Span>,
+}
+
+impl SubDiagnostic {
+    pub fn new(level: Level, message: String, span: Option<Span>) -> Self {
+        Self {
+            level,
+            message,
+            span,
+        }
+    }
+}
+
+/// A source location broken down into the file, the line, and the column. This is useful for
+/// showing diagnostics
+#[derive(Debug, Clone)]
+pub struct Loc {
+    /// The file that this location refers to
+    pub file: Rc<SourceFile>,
```

Comment:
Currently a lot of the machinery relies on being on a single thread. Interestingly that is also how Rustc is. In C we have to deal with the fact that many files will include the same header files, and we don't want to have to load those separately every time. Hence we will need some kind of hashing algorithm for the file path that allows us to see if we have already included a file. That doesn't have to be the same machinery as the SourceMap though, and it could be a separate entity for later in the compilation process.
As for being multi-threaded, it would also be nice to not include the same file twice if it is being compiled twice, but that impacts a lot of design decisions and I think I'd like to talk about that during one of our meetings. For now though, changing Rc<> to Arc<> will for now only slightly negatively impact performance rather than future proof the code, as it is already un-future-proofed.